buidl/message.py

```
@@ -0,0 +1,224 @@
+        # is to_sign always a segwit?
+        segwit=True
+    
+        # create unsigned to_sign transaction
+        to_sign_tx = Tx(version,tx_inputs,tx_outputs,locktime,network,segwit)
+        
+        if sig_bytes:
+            try:
+                stream = io.BytesIO(sig_bytes)
+                witness = Witness.parse(stream)
+                # Set the witness on the to_sign tx input
+                to_sign_tx.tx_ins[0].witness = witness
+            except:
+                # TODO: Fall back to legacy ...
+                print("Signature is niether an encoded witness or full transaction. Fall back to legacy")
+                return None
+                
+        return to_sign_tx
+
+# Test if sig_bytes can be decoded to a transaction
+# TODO: Is there a better way to test than this?
+def is_full_signature(sig_bytes):
+    try:
+        sig_stream = io.BytesIO(sig_bytes)
+
+        Tx.parse(sig_stream)
+    # TODO: more specific exception handling
+    except:
+        return False
+    return True
+        
+
+def sign_message(format: MessageSignatureFormat, private_key: PrivateKey, address: str, message: str):
+    
+    if (format != MessageSignatureFormat.LEGACY):
+        return sign_message_bip322(format,private_key,address,message)
+
+    script_pubkey = address_to_script_pubkey(address)
+    
+    if (not script_pubkey.is_p2pkh()):
+        raise ValueError("Address must be p2pkh for LEGACY signatures")
+        
+    # TODO: This legacy signing needs to produce a compact encoding of signature AND public key
+    # Needs implementing in the library I believe?
+    raise NotImplementedError("Legacy signing not yet implemented. Require compact encoding of sig and pubkey")
+    # b_msg = str_to_bytes(message)
+    # signature = private_key.sign_message(b_msg)
+    
+    # return base64_encode(signature.der())    
+    
+def sign_message_bip322(format: MessageSignatureFormat, private_key: PrivateKey, address: str, message: str):
```

Generally in sign message functions, there should be some sort of sanity check about whether it validates. If it doesn't then we can throw an error.

---

Isn't that what is implemented on line 195 - 197? Or did you have something else in mind